File: algorithm/dp_shortestPath.js
//this file is used to solve dijkstra algorithm
//to find nearest distance for each vertex i,j
//by dynamic programming, for example:
//min(A,G), if there are two points E and F pointing at G, then min(A,G) = min(min(A,E)+length(EG),min(A,F)+length(FG))
let vertex = 7;
let matrix = new Array(vertex);
for(let i = 0; i < vertex; i++){
	matrix[i] = new Array(vertex);
	matrix[i].fill(500);//we note 500 as maximum distance
	matrix[i][i] = 0;
}

console.log("initial matrix: ",matrix);

//add edge between u and v
//where 0 <= u,v <= vertex-1 && u != v
function addEdge(u,v,weight){
	matrix[u][v] = weight;
}

//here we have 7 vertexes, 0,1,2,3,4,5,6 and 8 edges
addEdge(0,1,1);
addEdge(0,2,3);
addEdge(1,2,4);
addEdge(1,4,6);
addEdge(1,3,3);
addEdge(4,6,4);
addEdge(2,3,5);
addEdge(3,6,2);

console.log("after adding edges, matrix is: ",matrix);

function dp(v, w,length){
	if(v == w) return 0;
	else{
		for(let i = 0; i < vertex; i++){
			if(i != w && matrix[i][w] != 500){
				length = Math.min(dp(v,i,length) + matrix[i][w],length);
			}
		}
		return length;
	}
}

console.log(dp(0,6,500));

File: algorithm/dp_shortestPath.cpp
//this file is used to solve dijkstra algorithm
//to find nearest distance for each vertex i,j
//by dynamic programming, for example:
//min(A,G), if there are two points E and F pointing at G, then min(A,G) = min(min(A,E)+length(EG),min(A,F)+length(FG))
#include <iostream>
#include <stdio.h>
#include <string>
#include<vector>
#include<stack>
#include<queue>
using namespace std;
class SparseGraph{
public:
	int n, m;//n nodes, m edges
	bool directed;//directed or non-directed
	vector<vector<int> > g;//2D vector

	//construct function
	SparseGraph(int n, bool directed){
		this->n = n;
		this->directed = directed;
		for(int i = 0; i < n; i++){
			g.push_back(vector<int>());
		}
		//establish matrix for storing distance
		for(int i = 0; i < n; i++){
			for(int j = 0; j < n; j++){
				if(i == j) g[i].push_back(0);
				else{
					g[i].push_back(500);//we use 500 to note maximum distance, or INT_MAX
				}
			}
		}
	}

	//deconstruct function
	~SparseGraph(){}

	//get node number
	int getV(){
		return n;
	}

	//get edge number
	int getE(){
		return m;
	}

	//add edge between two nodes
	void addEdge(int v, int w, int weight){
		g[v][w] = weight;//v -> w
		if(!directed){
			g[w][v] = weight;
		}//if non-directed, w -> v
		m++;//add an edge from v to w
	}

	void print(){
		for(int i = 0; i < g.size(); i++){
			cout<<"edge "<<i<<" : ";
			for(int j = 0; j < g[i].size(); j++){
				cout<<g[i][j]<<" ";
			}
			cout<<endl;
		}
	}

	int dp(int v, int w, int length){
		if(v == w) return 0;
		else{
			for(int i = 0; i < n; i++){
				if(i != w && g[i][w] != 500){
					length = min(dp(v,i,length) + g[i][w],length);
				}
			}
			return length;
		}
	}

};

int main(){
	
	int vertex,edge;
	cout<<"input vertex number"<<endl;
	cin>>vertex;
	cout<<"input edge number"<<endl;
	cin>>edge;
	SparseGraph myGraph(vertex,true);
	for(int i = 0; i < edge; i++){
		int from,to,weight;
		cout<<"add an edge and its weight: "<<endl;
		cin>>from>>to>>weight;
		myGraph.addEdge(from,to,weight);
	}

	myGraph.print();
	cout<<"input begin vertex and end vertex between 0 and "<<vertex-1<<endl;
	int begin,end;
	cin>>begin>>end;
	cout<<"shortest path is: "<<myGraph.dp(begin,end,500)<<endl;
    return 0;
}
